### DISCUSSION TITLE
Can we get some more guidance/docs on the recent development runtime change in 2.2?

### DISCUSSION THREAD

**Discussion by joostdecock** [mdx-js/mdx#2203] [Q&A] [answered]
> **Answer: https://github.com/orgs/mdx-js/discussions/2203#discussioncomment-4491981** (edited by wooorm)

The upgrade from 2.1 to 2.2 broke all of my MDX content with a frustrating error:

> _jsxDEV is not a function

Looking into what changed, it seems this is related to https://github.com/mdx-js/mdx/pull/2045 by @remcohaszing 

Now I don't pretend to understand all the intricate details, but what I do know is:

- This was a breaking change for me in a minor version bump. Not the end of the world, but just FYI, that's unexpected
- Setting `development: false` **does not** sidestap this issue for me. Short of `NODE_ENV=production` (or downgrading) I have yet to figure out how to resolve this.

I appreciate that this was debated as being a non-breaking change so maybe I'm doing something dumb which is why it breaks for me. But it would be nice to get some insight into what's going on and why `development: false` doesn't seem to fix it.

In my code, I'm doing what's documented under **MDX on demand**: https://mdxjs.com/guides/mdx-on-demand/

**Comment by joostdecock** [comment, marked as answer]
In case other people struggle with this, this is the current workaround I have put in place:

```js
import { run } from '@mdx-js/mdx'
import * as runtime from 'react/jsx-runtime'
// ...
await run(mdx, { ...runtime, jsxDEV: runtime.jsx })
```

But I'd like to understand rather than work around this issue.

**Reply by remcohaszing** [reply]
I’m sorry to see this caused some issues for you. This answer provides some information that was missing, so that helps troubleshooting the issue.

The JSX automatic production runtime imports from `react/jsx-runtime`. Development mode imports from `react/jsx-dev-runtime`. Usually this is handled by the build step, but when using `run()` you need to pass the JSX runtime yourself. I hadn’t thought of this exact situation, but I see why it’s breaking for you.

Disabling the `development` option **should** work. If it doesn’t, that’s a bug.

```js
import { run } from '@mdx-js/mdx'
import * as runtime from 'react/jsx-runtime'

const code = String(await compile(mdx, { development: false, outputFormat: 'function-body' }))
await run(code, { ...runtime })
```

If you do intend to use development mode, you should use this:

```js
import { run } from '@mdx-js/mdx'
import * as runtime from 'react/jsx-dev-runtime'

const code = String(await compile(mdx, { development: true, outputFormat: 'function-body' }))
await run(code, { ...runtime })
```

You should **not** pass `jsxDEV: runtime.jsx`. This may work by coincidence, but it’s incorrect and will lead to hard to troubleshoot bugs.

**Reply by joostdecock** [reply]
Thank you for taking the time to get back to my @remcohaszing 🙏 

After a night of sleep, I think I've figured it out, and in retrospect it of course all makes sense.
Here's what I currently understand to be the gist of it:

- The `develop: false` setting does in fact side-step/resolve the issue but **it must be passed to `compile()` and not to `run()`**. 
- This would not constitute a breaking change for scenarios where the runtime is auto-added. But it is a breaking change for the [MDX on demand](https://mdxjs.com/guides/mdx-on-demand/) behavior described.

I will submit a PR to add a note to the **MDX on-demand** documentation to make it easier for people in similar situtation to avoid getting bitten by this.

Thanks for all your hard work, and merry Christmas (if that's your thing) 🎄 🎅

**Reply by joostdecock** [reply]
https://github.com/mdx-js/mdx/pull/2204

**Reply by wooorm** [reply]
@remcohaszing can you update your code above to pass `development` to `compile` instead of `run`?
Other people might find this post and try to use your code, which is incorrect.

**Reply by colinsteidtmann** [reply]
Hey, I found this post after running into the same issue. Thanks y'all.

**Reply by ng-hai** [reply]
Thanks for the solution
